alibi_detect/utils/_types.py

```diff
@@ -2,8 +2,75 @@
 Defining types compatible with different Python versions and defining custom types.
 """
 import sys
+from typing import Any, Generic, Optional, Type, TypeVar
+import numpy as np
+from numpy.lib import NumpyVersion
+from pydantic.fields import ModelField
 
+# Literal for typing
 if sys.version_info >= (3, 8):
     from typing import Literal  # noqa
 else:
     from typing_extensions import Literal  # noqa
+
+# NumPy NDArray pydantic validator type
+# The code below is adapted from the great https://github.com/cheind/pydantic-numpy repo.
+# TODO - above repo also has some very clever functionality to allow passing the numpy array as a npy/npz file.
+#  This could be useful to explore in order to simplify our loading submodule.
+T = TypeVar("T", bound=np.generic)
+if NumpyVersion(np.__version__) < "1.22.0":
+    class NDArray(Generic[T], np.ndarray):
+        """
+        A Generic pydantic model to validate (and coerce) np.ndarray's.
+        """
+        @classmethod
+        def __get_validators__(cls):
+            yield cls.validate
+
+        @classmethod
+        def validate(cls, val: Any, field: ModelField) -> np.ndarray:
+            return _validate(cls, val, field)
+
+    class PotentialNDArray(Generic[T], np.ndarray):
+        """Like NDArray, but validation errors result in None."""
```

Comment:
What's the intended use case for this?

Reply:
It's literally the same as `NDArray`, but catches the `ValueError` and returns `None`. I've left it in there as thought there might be a use case for this in the future. Its not used at all atm though so we could just remove and add in again if needed?

Reply:
Yea I wasn't sure what the use case is for this, looking at the original docs seems it's used for loading files from disk? Perhaps best to remove if we don't use it.